### PR TITLE
Allow users to set resource requests in pipelines

### PR DESCRIPTION
**Pull request by msteffen**
and set a default resource request for pipelines that don't set one explicitly

### CODE REVIEW

File: src/server/pps/server/api_server.go [hunk 1]
```diff
@@ -2344,10 +2369,17 @@ func getJobOptions(kubeClient *kube.Client, jobInfo *persist.JobInfo, jobShimIma
...
+		memFootprint:       jobInfo.Resources.GetMemory(),
 	}, nil
 }
 
 func podSpec(options *jobOptions, jobID string, restartPolicy api.RestartPolicy) api.PodSpec {
+	mem, err := resource.ParseQuantity(options.memFootprint)
+	if err != nil {
+		mem = resource.MustParse(defaultMem)
+	}
+	cpu := resource.MustParse(fmt.Sprintf("%f", options.cpuFootprint))
```
Comment (sjezewski) [inline src/server/pps/server/api_server.go]: It seems like we are handling errors / defaults differently here for mem/cpu - is there a reason for that?

File: src/server/pps/server/api_server.go [hunk 2]
```diff
@@ -1273,6 +1277,25 @@ func setDefaultJobInputMethod(inputs []*ppsclient.JobInput) {
...
+const (
+	defaultCPU = 0.25
+	defaultMem = "250M"
+)
+
+// maybeDefaultResouceRequest returns 'res' if it's not nil, and a default
+// resource request it if is. This makes sure that pipeline/job workers are
+// given a baseline set of compute resources if none are specified in the
+// request.
+func maybeDefaultResourceRequest(res *ppsclient.Resources) *ppsclient.Resources {
```
Comment (sjezewski) [inline src/server/pps/server/api_server.go]: Personally I think this name is confusing ... I'd just name it `getResourceRequest` and its ok if it includes default behavior. But that's just me. This is a nitpick, not a dealbreaker.

File: src/server/pachyderm_test.go [hunk 3]
```diff
@@ -4468,6 +4468,76 @@ func TestListFileWithBlockSharding(t *testing.T) {
...
+	// Get info about the job pods from k8s & check for resources
+	kubeClient := getKubeClient(t)
+	retries := 10
+	for i := 0; i < retries; i++ {
+		podList, err := kubeClient.Pods(api.NamespaceDefault).List(api.ListOptions{
+			LabelSelector: labels.SelectorFromSet(
+				map[string]string{"app": jobInfos[0].Job.ID}),
+		})
+		// TODO(msteffen): Is there a more robust way to do this? We want to query
+		// kubernetes after the job is created and before the job is deleted.
```
Comment (sjezewski) [inline src/server/pachyderm_test.go]: Yea I don't think so (https://github.com/kubernetes/kubernetes/issues/1899). The only thing I could think of is maybe doing the equivalent of `kubectl logs job/someuuid -f` and block until input is received? But I'm not sure that'll work either. This seems ok for now.

File: src/server/pps/server/api_server.go [hunk 4]
```diff
@@ -2232,6 +2255,8 @@ type jobOptions struct {
 	volumes            []api.Volume
 	volumeMounts       []api.VolumeMount
 	imagePullSecrets   []api.LocalObjectReference
+	cpuFootprint       float32
+	memFootprint       string
```
Comment (sjezewski) [inline src/server/pps/server/api_server.go]: Cool. Yea we want this to be shown for inspect pipeline and inspect job. So I think the pachctl output commands need to be updated to print this information as well.

**Comment by sjezewski** [comment]
Ok, I've made the needed changes, now I need a review from you @msteffen

**Comment by sjezewski** [comment]
Oh, huh, whoops. I thought if I clicked 'approved' bcz of the changes I requested/made I could then request a review from you. Looks like no. So let's do it the old fashioned way and just use comments and a 'LGTM'

File: src/server/pps/pretty/pretty.go [hunk 5]
```diff
@@ -112,6 +112,9 @@ ParallelismSpec: {{.ParallelismSpec}}
 {{ if .Service }}Service:
 	{{ if .Service.InternalPort }}InternalPort: {{ .Service.InternalPort }} {{end}}
 	{{ if .Service.ExternalPort }}ExternalPort: {{ .Service.ExternalPort }} {{end}} {{end}}
+{{ if .Resources }}Resources:
+	CPU: {{ .Resources.Cpu }}
```
Comment (msteffen) [inline src/server/pps/pretty/pretty.go]: Do we need `{{ if .Resources.Cpu }}...{{end}}` around these?

File: src/server/pps/server/api_server.go (hunk 4 above)
Comment (msteffen) [inline src/server/pps/server/api_server.go]: Sounds good, though per our discussion, you implemented this :-)

**Comment by msteffen** [comment]
Per Sean and my discussion, this LGTM with Sean's changes. I think Sean is planning to maybe make a few more minor changes and submit it.

**Pull request merged**